moveit_core/cost_functions/src/cost_functions.cpp

```diff
@@ -0,0 +1,74 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2023, PickNik Inc.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of PickNik Inc. nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Sebastian Jahr */
+
+#include <moveit/cost_functions/cost_functions.hpp>
+
+namespace moveit
+{
+namespace cost_functions
+{
+[[nodiscard]] ::planning_interface::StateCostFn
+getClearanceCostFn(moveit::core::RobotState& robot_state, const std::string& group_name,
+                   const planning_scene::PlanningSceneConstPtr& planning_scene)
+{
+  // Create cost function
+  return [robot_state, group_name, planning_scene](const Eigen::VectorXd& state_vector) mutable {
```

Comment:
I think it is Ok because I capture references (e.g. `robot_state`) by value:
```
getClearanceCostFn(moveit::core::RobotState& robot_state, const std::string& group_name,
                   const planning_scene::PlanningSceneConstPtr& planning_scene)
{
  return [robot_state, group_name, planning_scene](const Eigen::VectorXd& state_vector)
}
```
Is that assumption correct?

Reply:
I think it is still making a copy, even if the original parameter is a reference (the pointer is ok though, because even if it makes a copy, it's a copy of the pointer). To verify I ran this minimal example:

```
#include <iostream>
#include <functional>

std::function<void()> create_function(int &p) {
   auto lambda=[p]() mutable {
        p += 50;
   };
   return lambda;
}

int main() {
        int a = 0;
        auto function = create_function(a);
        function();
        std::cout << a << std::endl;
}
```

This returns 0, i.e., the lambda works with a copy of `a`. But worse than that, the copy will be made when `create_function` is called, not when the actual lambda function is called, so it could be working on an old outdated state.

Reply:
Good to know! Thanks for the minimal example. I'll update the code